lxd/db/generate/README.md

```diff
@@ -0,0 +1,187 @@
+
+```
+
+This will initiate a call to `lxd-generate db mapper -t instances.mapper.go` on
+each call to the go generation directive `mapper`.
+
+### SQL Statement generation
+
+SQL generation supports the following SQL statement types:
+Type                                  | Description
+:---                                  | :----
+`objects`                             | Creates a basic SELECT statement of the form `SELECT <columns> FROM <table> ORDER BY <columns>`.
+`objects-by-<FIELD>-and-<FIELD>...`   | Parses a pre-existing SELECT statement of the form produced by`objects`, and appends a `WHERE` clause with the given fields located in the associated struct.
+`create`                              | Creates a basic INSERT statement of the form `INSERT INTO <table> VALUES`.
+`create-or-replace`                   | Creates a basic INSERT statement of the form `INSERT OR REPLACE INTO <table> VALUES`.
+`delete-by-<FIELD>-and-<FIELD>...`    | Creates a DELETE statement of the form `DELETE FROM <table> WHERE <constraint>` where the constraint is based on the given fields of the associated struct.
+`id`                                  | Creates a basic SELECT statement that returns just the internal ID of the table.
+`rename`                              | Creates an UPDATE statement that updates the primary key of a table: `UPDATE <table> SET <primary key> WHERE <primary key = ?>`.
+`update`                              | Creates an UPDATE statement of the form `UPDATE <table> SET <all columns> WHERE <primary key = ?>`.
+
+#### Examples
+
+```go
+//go:generate mapper stmt -e instance objects
+//go:generate mapper stmt -e instance objects-by-Name-and-Project
+//go:generate mapper -e instance create
+//go:generate mapper -e instance update
+//go:generate mapper -e instance delete-by-Name-and-Project
+```
+
+`//go:generate mapper stmt -e instance objects table=table_name`
+
+The `table` key can be used to override the generated table name for a specified one.
+
+#### Statement related Go tags
+
+There are several tags that can be added to fields of a struct that will be parsed by the `ast` package.
+
+Tag                         | Description
+:--                         | :----
+`coalesce=<value>`          | Generates a SQL coalesce function with the given value `coalesce(<column>, value)`.
+`order=yes`                 | Override the default `ORDER BY` columns with all fields specifying this tag.
+`join=<joinTable.column>`   | Applies a `JOIN` of the form `JOIN <joinTable> ON <table>.<joinTable_id> = <joinTable.id>`.
+`leftjoin=<table.column>`   | Applies a `LEFT JOIN` of the same form as a `JOIN`.
+`joinon=<column>`           | Overrides the default `JOIN ON` with the given column name, replacing `<joinTable_id>` above.
+`via=<otherTable>`          | Applies a nested join when included with a `join` tag.
+`primary=yes`               | Assigns the primary key for the table. Will default to `Name` if unspecified.
```

Comment:
This is one that regularly confuses me about the DB generator. As the DB generator doesn't generate the actual schema and thus has no control over the primary key. Instead it seems to influence the minimal set of arguments that should be supplied to specify a single row (but this doesn't have to be the primary key)? Is that correct?

Reply:
Yes, that's a much better description than just "assigns the primary key", I've now switched it to use that description.